File: src/main/java/net/thedragonteam/armorplus/worldgen/OreGen.java
/*
 * Copyright (c) TheDragonTeam 2016-2017.
 */

package net.thedragonteam.armorplus.worldgen;

import net.minecraft.block.state.pattern.BlockMatcher;
import net.minecraft.init.Blocks;
import net.minecraft.world.World;
import net.minecraft.world.chunk.IChunkProvider;
import net.minecraft.world.gen.IChunkGenerator;
import net.minecraft.world.gen.feature.WorldGenMinable;
import net.minecraft.world.gen.feature.WorldGenerator;
import net.minecraftforge.fml.common.IWorldGenerator;
import net.thedragonteam.armorplus.ModConfig.WorldGenConfig.OreLavaCrystal.DimensionOre;
import net.thedragonteam.armorplus.registry.ModBlocks;
import net.thedragonteam.armorplus.util.WorldGenUtils;

import java.util.Random;

import static net.thedragonteam.armorplus.ModConfig.WorldGenConfig.lava_crystal;

/**
 * @author Sokratis Fotkatzikis - TheDragonTeam
 **/
public class OreGen implements IWorldGenerator {

    private WorldGenerator overworldGenerator = new WorldGenMinable(ModBlocks.oreLavaCrystal.getDefaultState(), lava_crystal.overworld.veinAmount);
    private WorldGenerator theEndGenerator = new WorldGenMinable(ModBlocks.oreLavaCrystal.getDefaultState(), lava_crystal.the_end.veinAmount, BlockMatcher.forBlock(Blocks.END_STONE));
    private WorldGenerator theNetherGenerator = new WorldGenMinable(ModBlocks.oreLavaCrystal.getDefaultState(), lava_crystal.the_nether.veinAmount, BlockMatcher.forBlock(Blocks.NETHERRACK));

    @Override
    public void generate(Random random, int chunkX, int chunkZ, World world, IChunkGenerator chunkGenerator, IChunkProvider chunkProvider) {
        int i = world.provider.getDimension();
        if (i == 0) {
            if (lava_crystal.overworld.enable) {
                runGenerator(overworldGenerator, world, random, chunkX, chunkZ, lava_crystal.overworld);
            }
        } else if (i == 1) {
            if (lava_crystal.the_end.enable) {
                runGenerator(theEndGenerator, world, random, chunkX, chunkZ, lava_crystal.the_end);
            }
        } else if (i == -1) {
            if (lava_crystal.the_nether.enable) {
                runGenerator(theNetherGenerator, world, random, chunkX, chunkZ, lava_crystal.the_nether);
            }
        }
    }

    private void runGenerator(WorldGenerator generator, World world, Random random, int chunkX, int chunkZ, DimensionOre dimensionOre) {
        WorldGenUtils.runGenerator(generator, world, random, chunkX, chunkZ, dimensionOre.rarity, dimensionOre.minYSpawn, dimensionOre.maxYSpawn);
    }
}

File: src/main/java/net/thedragonteam/armorplus/worldgen/OreGen.kt
/*
 * Copyright (c) TheDragonTeam 2016-2017.
 */

package net.thedragonteam.armorplus.worldgen

import net.minecraft.block.state.pattern.BlockMatcher
import net.minecraft.init.Blocks
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraft.world.chunk.IChunkGenerator
import net.minecraft.world.chunk.IChunkProvider
import net.minecraft.world.gen.feature.WorldGenMinable
import net.minecraft.world.gen.feature.WorldGenerator
import net.minecraftforge.fml.common.IWorldGenerator
import net.thedragonteam.armorplus.APConfig.*
import net.thedragonteam.armorplus.registry.ModBlocks
import java.util.*

/**
 * net.thedragonteam.armorplus.worldgen
 * ArmorPlus created by sokratis12GR
 * - TheDragonTeam
 */
class OreGen : IWorldGenerator {

    private val lavaCrystalOverworldGenerator: WorldGenerator = WorldGenMinable(ModBlocks.blockLavaCrystal.defaultState, lavaCrystalOverworldVeinAmountWorking)
    private val lavaCrystalTheEndGenerator: WorldGenerator = WorldGenMinable(ModBlocks.blockLavaCrystal.defaultState, lavaCrystalTheEndVeinAmount, BlockMatcher.forBlock(Blocks.END_STONE))
    private val lavaCrystalTheNetherGenerator: WorldGenerator = WorldGenMinable(ModBlocks.blockLavaCrystal.defaultState, lavaCrystalTheNetherVeinAmount, BlockMatcher.forBlock(Blocks.NETHERRACK))

    override fun generate(random: Random, chunkX: Int, chunkZ: Int, world: World, chunkGenerator: IChunkGenerator, chunkProvider: IChunkProvider) {
        when (world.provider.dimension) {
        //The End
            1 -> if (enableLavaCrystalTheEndGen) {
                this.runGenerator(lavaCrystalTheEndGenerator, world, random, chunkX, chunkZ, lavaCrystalTheEndRarity, lavaCrystalTheEndMinYSpawn, lavaCrystalTheEndMaxYSpawn)
            }
        //Overworld Dimension
            0 -> if (enableLavaCrystalOverworldGen) {
                this.runGenerator(lavaCrystalOverworldGenerator, world, random, chunkX, chunkZ, lavaCrystalOverworldRarityWorkingOne, lavaCrystalOverworldMinYSpawn, lavaCrystalOverworldMaxYSpawn)
            }
        //The Nether
            -1 -> if (enableLavaCrystalTheNetherGen) {
                this.runGenerator(lavaCrystalTheNetherGenerator, world, random, chunkX, chunkZ, lavaCrystalTheNetherRarity, lavaCrystalTheNetherMinYSpawn, lavaCrystalTheNetherMaxYSpawn)
            }
        }
    }

    private fun runGenerator(generator: WorldGenerator, world: World, rand: Random, chunk_X: Int, chunk_Z: Int, chancesToSpawn: Int, minHeight: Int, maxHeight: Int) {
        assert(!(minHeight < 0 || maxHeight > 256 || minHeight > maxHeight)) { "Illegal Height Arguments for WorldGenerator" }
        val heightDiff = maxHeight - minHeight + 1
        for (i in 0 .. chancesToSpawn - 1) {
            val x = chunk_X * 16 + rand.nextInt(16)
            val y = minHeight + rand.nextInt(heightDiff)
            val z = chunk_Z * 16 + rand.nextInt(16)
            val orePos = BlockPos(x, y, z)
            generator.generate(world, rand, orePos)
        }
    }
}